### PR TITLE
Feature: set/unset container attribute

### PR DESCRIPTION
**Pull request by Zorin95670**
:information_source:  Don't forget to modify the changelog.md before merging this branch.

### CODE REVIEW

**Review by Hundraw — DISMISSED** [review]

**Review by Hundraw — DISMISSED** [review]

File: src/models/Component.js [hunk 1]
```diff
@@ -66,20 +67,51 @@ class Component extends FileInformation {
...
+        type: 'String',
+        definition: attributeDefinition,
+      }));
+    }
+  }
+
+  /**
+   * Remove all attributes referring to the container.
+   */
+  unsetAllContainerAttribute() {
```
Comment (sohzaz) [inline src/models/Component.js:103]: Maybe rename to unsetAllReferences?

File: tests/unit/models/Component.spec.js [hunk 2]
```diff
@@ -49,44 +51,213 @@ describe('Test class: Component', () => {
...
+        });
+        component.setContainerAttribute(new Component({
+          id: 'containerId',
+          definition: new ComponentDefinition({
+            type: 'container',
+          }),
+        }));
+
+        expect(component.attributes.length).toEqual(1);
+        expect(component.attributes[0].name).toEqual('container');
```
Comment (sohzaz) [inline tests/unit/models/Component.spec.js:124]: ``
expect(component.attributes[0]).toMatchObject({
name: 'container', value:'containerId',
})
``

File: tests/unit/models/Component.spec.js [hunk 3]
```diff
@@ -49,44 +51,213 @@ describe('Test class: Component', () => {
...
+            }),
+          ],
+        });
+        component.unsetContainerAttribute(new Component({
+          id: 'containerId',
+          definition: new ComponentDefinition({
+            type: 'container',
+          }),
+        }));
+
```
Comment (sohzaz) [inline tests/unit/models/Component.spec.js:194]: explicitly test that the correct attribute has been removed

File: tests/unit/models/Component.spec.js [hunk 4]
```diff
@@ -49,44 +51,213 @@ describe('Test class: Component', () => {
...
+        });
+        component.setContainerAttribute(new Component({
+          id: 'containerId',
+          definition: new ComponentDefinition({
+            type: 'container',
+          }),
+        }));
+
+        expect(component.attributes.length).toEqual(1);
+        expect(component.attributes[0].name).toEqual('container');
```
Comment (sohzaz) [inline tests/unit/models/Component.spec.js:88]: ``
expect(component.attributes[0]).toMatchObject({
name: 'container', value:'containerId',
})
``

File: tests/unit/models/Component.spec.js [hunk 5]
```diff
@@ -49,44 +51,213 @@ describe('Test class: Component', () => {
...
+            }),
+          ],
+        });
+        component.unsetContainerAttribute(new Component({
+          id: 'containerId',
+          definition: new ComponentDefinition({
+            type: 'container',
+          }),
+        }));
+
```
Comment (sohzaz) [inline tests/unit/models/Component.spec.js:228]: explicitly test that the correct attribute has been removed

File: tests/unit/models/Component.spec.js [hunk 6]
```diff
@@ -49,44 +51,213 @@ describe('Test class: Component', () => {
...
+              definition: new ComponentAttributeDefinition({
+                name: 'container',
+                type: 'String',
+              }),
+              value: 'test',
+            }),
+          ],
+        });
+        component.unsetAllContainerAttribute();
+
```
Comment (sohzaz) [inline tests/unit/models/Component.spec.js:258]: explicitly test that all the correct attributes have been removed

File: tests/unit/models/Component.spec.js [hunk 7]
```diff
@@ -49,44 +51,213 @@ describe('Test class: Component', () => {
...
-        expect(parent.children.length).toEqual(1);
-        expect(parent.children[0]).not.toBeNull();
-        expect(parent.children[0].id).toEqual('2');
-
-        parent.addChild(child2);
-        expect(parent.children.length).toEqual(2);
-        expect(parent.children[1]).not.toBeNull();
-        expect(parent.children[1].id).toEqual('3');
+    describe('Test method: setContainerAttribute', () => {
+      it('should not set attribute if there are no container attribute definition', () => {
```
Comment (k-y0u) [inline tests/unit/models/Component.spec.js:55]: `there are` -> `there is`

File: tests/unit/models/Component.spec.js [hunk 8]
```diff
@@ -49,44 +51,213 @@ describe('Test class: Component', () => {
...
-        const parent = new Component({ id: '1', definition: componentDefinition });
-        const child = new Component({ id: '2', definition: { parentTypes: ['parent1'] } });
-        parent.addChild(child);
-        expect(parent.children.length).toEqual(1);
-        parent.addChild(child);
-        expect(parent.children.length).toEqual(1);
+    });
+
+    describe('Test method: unsetContainerAttribute', () => {
+      it('Should remove attribute if exists', () => {
```
Comment (k-y0u) [inline tests/unit/models/Component.spec.js:131]: `attribute if exists`
-> `attribute if it exists`
-> `existing attribute`

File: tests/unit/models/Component.spec.js [hunk 9]
```diff
@@ -49,44 +51,213 @@ describe('Test class: Component', () => {
...
+          definition: new ComponentDefinition({
+            type: 'container',
+          }),
+        }));
+
+        expect(component.attributes.length)
+          .toEqual(0);
+      });
+
+      it('Should do nothing if attribute not exists', () => {
```
Comment (k-y0u) [inline tests/unit/models/Component.spec.js:165]: same -> does

File: tests/unit/models/Component.spec.js [hunk 10]
```diff
@@ -49,44 +51,213 @@ describe('Test class: Component', () => {
...
+          definition: new ComponentDefinition({
+            type: 'container',
+          }),
+        }));
+
+        expect(component.attributes.length)
+          .toEqual(2);
+      });
+
+      it('Should do nothing if attribute not exists', () => {
```
Comment (k-y0u) [inline tests/unit/models/Component.spec.js:199]: same name as above...

File: tests/unit/models/Component.spec.js [hunk 11]
```diff
@@ -49,44 +51,213 @@ describe('Test class: Component', () => {
...
+          }),
+        }));
+
+        expect(component.attributes.length)
+          .toEqual(2);
+      });
+    });
+
+    describe('Test method: unsetAllContainerAttribute', () => {
+      it('Should remove all container attribute', () => {
```
Comment (k-y0u) [inline tests/unit/models/Component.spec.js:235]: all -> attributeS

**Review by k-y0u — CHANGES_REQUESTED** [review]

File: src/models/Component.js (hunk 1 above)
Comment (Zorin95670) [reply src/models/Component.js:103]: done

File: tests/unit/models/Component.spec.js (hunk 2 above)
Comment (Zorin95670) [reply tests/unit/models/Component.spec.js:124]: done

File: tests/unit/models/Component.spec.js (hunk 3 above)
Comment (Zorin95670) [reply tests/unit/models/Component.spec.js:194]: done

File: tests/unit/models/Component.spec.js (hunk 4 above)
Comment (Zorin95670) [reply tests/unit/models/Component.spec.js:88]: done

File: tests/unit/models/Component.spec.js (hunk 5 above)
Comment (Zorin95670) [reply tests/unit/models/Component.spec.js:228]: done

File: tests/unit/models/Component.spec.js (hunk 6 above)
Comment (Zorin95670) [reply tests/unit/models/Component.spec.js:258]: done

File: src/models/Component.js [hunk 12]
```diff
@@ -66,20 +67,51 @@ class Component extends FileInformation {
...
+    this.attributes = this.attributes.filter(({ definition }) => definition.type !== 'Reference');
+  }
+
+  /**
+   * Remove attribute referring to the container.
+   * @param {Component} container - Container.
+   */
+  unsetContainerAttribute(container) {
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/models/Component.js:111]: done

File: tests/unit/models/Component.spec.js [hunk 13]
```diff
@@ -49,44 +51,213 @@ describe('Test class: Component', () => {
...
+        expect(component.attributes).toEqual([]);
       });
-      it('Check not passing child if isContainer is false', () => {
-        const componentDefinition = new ComponentDefinition({ isContainer: false, type: 'parent1' });
-        const parent = new Component({ id: '1', definition: componentDefinition });
-        const child = new Component({ id: '2', definition: { parentTypes: ['parent1'] } });
-        parent.addChild(child);
-        expect(parent.children.length).toEqual(0);
+
+      it('should create attribute if container attribute not exits', () => {
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply tests/unit/models/Component.spec.js:67]: done

File: tests/unit/models/Component.spec.js (hunk 7 above)
Comment (Zorin95670) [reply tests/unit/models/Component.spec.js:55]: done

File: tests/unit/models/Component.spec.js (hunk 8 above)
Comment (Zorin95670) [reply tests/unit/models/Component.spec.js:131]: done

File: tests/unit/models/Component.spec.js (hunk 11 above)
Comment (Zorin95670) [reply tests/unit/models/Component.spec.js:235]: done

File: tests/unit/models/Component.spec.js (hunk 9 above)
Comment (Zorin95670) [reply tests/unit/models/Component.spec.js:165]: done

**Review by sohzaz — APPROVED** [review]

**Review by k-y0u — APPROVED** [review]

**Review by sohzaz — APPROVED** [review]

**Pull request merged**